### PR TITLE
Deep Exponential Family Example

### PR DESCRIPTION
**Pull request by matthew-mcateer**
Added Jupyter Notebook Detailing how to implement a Deep Exponential Family. This was based on an example from the edward1 library.

Original content [this Repository](https://github.com/blei-lab/edward/blob/master/examples/deep_exponential_family.py) and [this paper](http://www.cs.toronto.edu/~lcharlin/papers/def_aistats.pdf), created by [the Blei Lab](http://www.cs.columbia.edu/~blei/)

### CODE REVIEW

File: tensorflow_probability/examples/jupyter_notebooks/Deep_Exponential_Family_TFP.ipynb [hunk 1]
```diff
@@ -0,0 +1,1904 @@
...
+        "colab_type": "code",
+        "colab": {}
+      },
+      "cell_type": "code",
+      "source": [
+        "def trainable_positive_deterministic(shape, min_loc=1e-3, name=None):\n",
+        "    \"\"\"Learnable Deterministic distribution over positive reals.\"\"\"\n",
+        "    with tf.variable_scope(None, default_name=\"trainable_positive_deterministic\"):\n",
+        "        unconstrained_loc = tf.get_variable(\"unconstrained_loc\", shape)\n",
+        "        loc = tf.maximum(tf.nn.softplus(unconstrained_loc), min_loc)\n",
```
Comment (brianwa84) [inline tensorflow_probability/examples/jupyter_notebooks/Deep_Exponential_Family_TFP.ipynb]: maximum is not differentiable to the non-max element (i.e. once a value falls below threshold it can never come back). could we just use `softplus(unconstrained)+eps`?

File: tensorflow_probability/examples/jupyter_notebooks/Deep_Exponential_Family_TFP.ipynb [hunk 2]
```diff
@@ -0,0 +1,1904 @@
...
+      },
+      "cell_type": "code",
+      "source": [
+        "def trainable_gamma(shape, min_concentration=1e-3, min_scale=1e-5, name=None):\n",
+        "    \"\"\"Learnable Gamma via concentration and scale parameterization.\"\"\"\n",
+        "    with tf.variable_scope(None, default_name=\"trainable_gamma\"):\n",
+        "        unconstrained_concentration = tf.get_variable(\n",
+        "            \"unconstrained_concentration\", shape,\n",
+        "            initializer=tf.random_normal_initializer(mean=0.5, stddev=0.1))\n",
+        "        unconstrained_scale = tf.get_variable(\n",
```
Comment (brianwa84) [inline tensorflow_probability/examples/jupyter_notebooks/Deep_Exponential_Family_TFP.ipynb]: the v2 way would be to explicitly init a Variable instance
`tf.compat.v2.Variable(..)`

File: tensorflow_probability/examples/jupyter_notebooks/Deep_Exponential_Family_TFP.ipynb [hunk 3]
```diff
@@ -0,0 +1,1904 @@
...
+        "                     tf.fill([layer_sizes[0], feature_size], 0.3), \n",
+        "                     name=\"w0_prior\")\n",
+        "z2_prior = tfd.Gamma(tf.fill([data_size, layer_sizes[2]], 0.1), \n",
+        "                     tf.fill([data_size, layer_sizes[2]], 0.1), \n",
+        "                     name=\"z2_prior\")\n",
+        "z1_prior = tfd.Gamma(shape, \n",
+        "                     shape / tf.matmul(z2_prior.sample(), w2_prior.sample()), \n",
+        "                     name=\"z1_prior\")\n",
+        "z0_prior = tfd.Gamma(shape, \n",
+        "                     shape / tf.matmul(z1_prior.sample(), w1_prior.sample()), \n",
```
Comment (brianwa84) [inline tensorflow_probability/examples/jupyter_notebooks/Deep_Exponential_Family_TFP.ipynb]: you want two distinct samples here?

File: tensorflow_probability/examples/jupyter_notebooks/Deep_Exponential_Family_TFP.ipynb [hunk 4]
```diff
@@ -0,0 +1,1904 @@
...
+            "Topic 3: maintain shortly accomplished specifies lausanne black rise invariance instead shrink\n",
+            "Topic 4: regressors saw orders phrase star observes sapiro impression site morgan\n",
+            "Topic 5: cauchy separation tenenbaum interscience agglomerative together malicious environment collaborative berkeley\n",
+            "Topic 6: describing syst completed opposed capable lambda tendency riding restaurant hall\n",
+            "Topic 7: treated location lemmas segments connected static imposed lane sphere symbol\n",
+            "Topic 8: speaker darpa refinement adversarial savings altun vary may kmax shamir\n",
+            "Topic 9: overhead platform ibp evaluations ronald clearly extensions dirichlet jacobian interna\n",
+            "Step: 7000 Loss: -2989526016.000 (0.192 sec)\n",
+            "Negative log-likelihood <= 2989526016.000\n",
+            "Perplexity <= inf\n"
```
Comment (brianwa84) [inline tensorflow_probability/examples/jupyter_notebooks/Deep_Exponential_Family_TFP.ipynb]: uh oh? and the following samples reflect the failure

File: tensorflow_probability/examples/jupyter_notebooks/Deep_Exponential_Family_TFP.ipynb [hunk 5]
```diff
@@ -0,0 +1,1904 @@
...
+        "        duration = time.time() - start_time\n",
+        "        print(\"Step: {:>3d} Loss: {:.3f} ({:.3f} sec)\".format(\n",
+        "              step, elbo_value, duration))\n",
+        "        summary_str = evaluate(summary)\n",
+        "        summary_writer.add_summary(summary_str, step)\n",
+        "        summary_writer.flush()\n",
+        "\n",
+        "        # Compute perplexity of the full data set. The model's negative\n",
+        "        # log-likelihood of data is upper bounded by the variational objective.\n",
+        "        negative_log_likelihood = -elbo_value\n",
```
Comment (brianwa84) [inline tensorflow_probability/examples/jupyter_notebooks/Deep_Exponential_Family_TFP.ipynb]: not much point in printing both the positive and negative likelihoods.

I think this cell needs to succeed more often than it fails to check it in. Can we lower the learning rate or adjust the prior gamma params?

**Comment by brianwa84** [comment]
As with the other PR (probably all of them) it'd be nice to be demonstrating TF2 instead of TF1, but imo it's fair to leave that to later.
